File: cdp-agentkit-core/cdp_agentkit_core/actions/transfer.py
from collections.abc import Callable

from cdp import Wallet
from pydantic import BaseModel, Field

from cdp_agentkit_core.actions import CdpAction

TRANSFER_PROMPT = """
This tool will transfer an asset from the wallet to another onchain address.

It takes the following inputs:
- amount: The amount to transfer
- assetId: The asset ID to transfer
- destination: Where to send the funds (can be an onchain address, ENS 'example.eth', or Basename 'example.base.eth')
- gasless: Whether to do a gasless transfer

Important notes:
- Gasless transfers are only available on base-sepolia and base-mainnet (base) networks for 'usdc' asset
- Always use gasless transfers when available
- Always use asset ID 'usdc' when transferring USDC
- Ensure sufficient balance of the input asset before transferring
- When sending native assets (e.g. 'eth' on base-mainnet), ensure there is sufficient balance for the transfer itself AND the gas cost of this transfer
"""

class TransferInput(BaseModel):
    """Input argument schema for transfer action."""

    amount: str = Field(
        ..., description="The amount of the asset to transfer, e.g. `15`, `0.000001`"
    )
    asset_id: str = Field(
        ...,
        description="The asset ID to transfer, e.g. `eth`, `0x036CbD53842c5426634e7929541eC2318f3dCF7e`",
    )
    destination: str = Field(
        ...,
        description="The destination to transfer the funds, e.g. `0x58dBecc0894Ab4C24F98a0e684c989eD07e4e027`, `example.eth`, `example.base.eth`",
    )
    gasless: bool = Field(
        default=False,
        description="whether to do a gasless transfer (gasless is available on Base Sepolia and Mainnet for USDC) Always do the gasless option when it is available.",
    )


def transfer(
    wallet: Wallet, amount: str, asset_id: str, destination: str, gasless: bool = False
) -> str:
    """Transfer a specified amount of an asset to a destination onchain. USDC Transfers on Base Sepolia and Mainnet can be gasless. Always use the gasless option when available.

    Args:
        wallet (Wallet): The wallet to transfer the asset from.
        amount (str): The amount of the asset to transfer, e.g. `15`, `0.000001`.
        asset_id (str): The asset ID to transfer (e.g., "eth", "usdc", or a valid contract address like "0x036CbD53842c5426634e7929541eC2318f3dCF7e").
        destination (str): The destination to transfer the funds (e.g. `0x58dBecc0894Ab4C24F98a0e684c989eD07e4e027`, `example.eth`, `example.base.eth`).
        gasless (bool): Whether to send a gasless transfer (Defaults to False.).

    Returns:
        str: A message containing the transfer details.

    """
    try:
        transfer_result = wallet.transfer(
            amount=amount, asset_id=asset_id, destination=destination, gasless=gasless
        ).wait()
    except Exception as e:
        return f"Error transferring the asset {e!s}"

    return f"Transferred {amount} of {asset_id} to {destination}.\nTransaction hash for the transfer: {transfer_result.transaction_hash}\nTransaction link for the transfer: {transfer_result.transaction_link}"


class TransferAction(CdpAction):
    """Transfer action."""

    name: str = "transfer"
    description: str = TRANSFER_PROMPT
    args_schema: type[BaseModel] | None = TransferInput
    func: Callable[..., str] = transfer

File: cdp-agentkit-core/typescript/src/actions/cdp/transfer.ts
import { CdpAction } from "./cdp_action";
import { Wallet, Amount } from "@coinbase/coinbase-sdk";
import { z } from "zod";

const TRANSFER_PROMPT = `
This tool will transfer an asset from the wallet to another onchain address.

It takes the following inputs:
- amount: The amount to transfer
- assetId: The asset ID to transfer
- destination: Where to send the funds (can be an onchain address, ENS 'example.eth', or Basename 'example.base.eth')
- gasless: Whether to do a gasless transfer

Important notes:
- Gasless transfers are only available on base-sepolia and base-mainnet (base) networks for 'usdc' asset
- Always use gasless transfers when available
- Always use asset ID 'usdc' when transferring USDC
- Ensure sufficient balance of the input asset before transferring
- When sending native assets (e.g. 'eth' on base-mainnet), ensure there is sufficient balance for the transfer itself AND the gas cost of this transfer
`;

/**
 * Input schema for transfer action.
 */
export const TransferInput = z
  .object({
    amount: z.custom<Amount>().describe("The amount of the asset to transfer"),
    assetId: z.string().describe("The asset ID to transfer"),
    destination: z.string().describe("The destination to transfer the funds"),
    gasless: z.boolean().default(false).describe("Whether to do a gasless transfer"),
  })
  .strip()
  .describe("Instructions for transferring assets");

/**
 * Transfers a specified amount of an asset to a destination onchain.
 *
 * @param wallet - The wallet to transfer the asset from.
 * @param args - The input arguments for the action.
 * @returns A message containing the transfer details.
 */
export async function transfer(
  wallet: Wallet,
  args: z.infer<typeof TransferInput>,
): Promise<string> {
  try {
    const transferResult = await wallet.createTransfer({
      amount: args.amount,
      assetId: args.assetId,
      destination: args.destination,
      gasless: args.gasless,
    });

    const result = await transferResult.wait();

    return `Transferred ${args.amount} of ${args.assetId} to ${args.destination}.\nTransaction hash for the transfer: ${result.getTransactionHash()}\nTransaction link for the transfer: ${result.getTransactionLink()}`;
  } catch (error) {
    return `Error transferring the asset: ${error}`;
  }
}

/**
 * Transfer action.
 */
export class TransferAction implements CdpAction<typeof TransferInput> {
  public name = "transfer";
  public description = TRANSFER_PROMPT;
  public argsSchema = TransferInput;
  public func = transfer;
}
